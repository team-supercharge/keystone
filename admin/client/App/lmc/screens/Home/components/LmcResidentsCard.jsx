import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import {
    Button,
	GlyphButton,
	ResponsiveText,
} from '../../../../elemental';
import { Link } from 'react-router';
import LmcProfileLink from '../../../components/LmcProfileLink.jsx';


class LmcResidentsCard extends Component {

    constructor (props) {
        super(props);
        this.state = {
            max_residents_displayed: MAX_RESIDENTS_DISPLAYED_INIT,
        };
        this.showMore = this.showMore.bind(this);
        this.renderResidents = this.renderResidents.bind(this);
    }

    showMore () {
        this.setState({
            max_residents_displayed: this.state.max_residents_displayed + 20, // + 1 row
        });
    }

    renderResidentCount (count, home) {
        return (
            <p>
                { home
                    ? <span>{ home } has </span>
                    : 'There are ' }
                <strong>{ count } active residents</strong>
            </p>
        );
    }

    renderResidents (activeResidents) {
        const { max_residents_displayed } = this.state;
        const hiddenResidents = activeResidents.length > max_residents_displayed;
        const home = activeResidents[0].home;
        return (
            <div>
                { home && activeResidents.length > 10
                    ? this.renderResidentCount(activeResidents.length, home)
                    : null }
                <div className="lmc-flex-grid">
                    { _.take(activeResidents, hiddenResidents ? max_residents_displayed - 1 : max_residents_displayed)
                        .map((row, index) =>
                            <LmcProfileLink
                                to={`${ Keystone.adminPath }/residents/${ row.id }`}
                                key={row.id}
                                name={row.name}
                                picture={row.picture}
                            />) }
                    { hiddenResidents
                        ? <a className="lmc-more-link" onClick={this.showMore}>More...</a>
                        : null }
                </div>
            </div>
        );
    }

    renderNoResidents () {
        return (
            <p>
                { NO_RESIDENTS_INFO }
            </p>
        );
    }

    renderNoActiveResidents () {
        return (
            <p>
                There are no active residents.
            </p>
        );
    }

    render () {
        const { residents } = this.props;
        let activeResidents;
        if (residents && residents.length) {
            activeResidents = _.chain(residents)
                .filter(res => res.status === 'active')
                .sortBy('name')
                .value();
        }

        const onClick = () => {
            this.props.onCreate('Resident');
        };

        return (
            <div>
                <h2 className="lmc-card-title">
                    { TITLE }
                </h2>
                <div className="lmc-card">
                    <div className="lmc-card-body">
                        { activeResidents && activeResidents.length
                                ? this.renderResidents(activeResidents)
                                : this.renderNoResidents() }
                    </div>
                    <div className="lmc-card-footer">
                        <div className="lmc-flex-container">
                            <div style={{ maxWidth: 160 }} id="intro-js-step-add-residents">
                                <GlyphButton
                                    block
                                    color="success"
                                    glyph="plus"
                                    onClick={onClick}
                                    position="left"
                                    title={ADD_RESIDENTS_BUTTON_TEXT}
                                >
                                    <ResponsiveText
                                        visibleSM={ADD_RESIDENTS_BUTTON_TEXT}
                                        visibleMD={ADD_RESIDENTS_BUTTON_TEXT}
                                        visibleLG={ADD_RESIDENTS_BUTTON_TEXT}
                                    />
                                </GlyphButton>
                            </div>
                            <div style={{ maxWidth: 95 }}>
                                <Link to={`${Keystone.adminPath}/residents`}>
                                    <Button color="default">
                                        <span style={{ opacity: 0.6 }}>
                                            { VIEW_BUTTON_TEXT }
                                        </span>
                                    </Button>
                                </Link>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}

// const styles = {
//     title: {
//         opacity: 0.8,
//     }
// }

LmcResidentsCard.propTypes = {
    onCreate: PropTypes.func.isRequired,
    residents: PropTypes.array.isRequired,
};

const MAX_RESIDENTS_DISPLAYED_INIT = 10;
const TITLE = 'Residents';
const VIEW_BUTTON_TEXT = 'View All';
const ADD_RESIDENTS_BUTTON_TEXT = 'Add a Resident';
// const NO_ACTIVE_RESIDENTS = ``;
const NO_RESIDENTS_INFO = `You have not added any residents yet. 
    Youâ€™ll need to add them here before anything appears in the Carer App`;

export default LmcResidentsCard;
