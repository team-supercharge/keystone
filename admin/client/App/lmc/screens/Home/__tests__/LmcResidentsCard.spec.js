import React from 'react'
import { shallow } from 'enzyme'
import LmcResidentsCard from '../components/LmcResidentsCard'

describe('LmcResidentsCard', () => {
    let wrapper
    let savedKeystone
    let residents

    beforeAll(() => {
        savedKeystone = global.Keystone
        global.Keystone = { adminPath: 'TestPath' }
    })

    beforeEach(() => {
        residents = [
            {
                name: 'TestName',
                id: 1,
                status: 'active'
            }
        ]
        wrapper = shallow(
            <LmcResidentsCard
                onCreate={() => {}}
                residents={residents}
            />
        )
    })

    test('should render correct snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('should have a link to the correct resident profile', () => {
        const profileLink = wrapper.find('LmcProfileLink')
        expect(profileLink.props().name).toEqual('TestName')
        expect(profileLink.props().to).toEqual('TestPath/residents/1')
    })

    afterAll(() => {
        global.Keystone = savedKeystone
    })
})