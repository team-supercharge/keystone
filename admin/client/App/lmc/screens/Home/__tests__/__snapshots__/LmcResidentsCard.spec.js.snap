// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LmcResidentsCard should render correct snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <h2
      className="lmc-card-title"
    >
      Residents
    </h2>
    <div
      className="lmc-card"
    >
      <div
        className="lmc-card-body"
      >
        <div>
          <div
            className="lmc-flex-grid"
          >
            <LmcProfileLink
              name="TestName"
              to="TestPath/residents/1"
            />
          </div>
        </div>
      </div>
      <div
        className="lmc-card-footer"
      >
        <div
          className="lmc-flex-container"
        >
          <div
            id="intro-js-step-add-residents"
            style={
              Object {
                "maxWidth": 160,
              }
            }
          >
            <GlyphButton
              block={true}
              color="success"
              glyph="plus"
              glyphStyle={Object {}}
              onClick={[Function]}
              position="left"
              title="Add a Resident"
            >
              <ResponsiveText
                component="span"
                visibleLG="Add a Resident"
                visibleMD="Add a Resident"
                visibleSM="Add a Resident"
              />
            </GlyphButton>
          </div>
          <div
            style={
              Object {
                "maxWidth": 95,
              }
            }
          >
            <Link
              onlyActiveOnIndex={false}
              style={Object {}}
              to="TestPath/residents"
            >
              <Button
                aphroditeStyles={Array []}
                color="default"
                variant="fill"
              >
                <span
                  style={
                    Object {
                      "opacity": 0.6,
                    }
                  }
                >
                  View All
                </span>
              </Button>
            </Link>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "nodes": Array [
    <div>
      <h2
        className="lmc-card-title"
      >
        Residents
      </h2>
      <div
        className="lmc-card"
      >
        <div
          className="lmc-card-body"
        >
          <div>
            <div
              className="lmc-flex-grid"
            >
              <LmcProfileLink
                name="TestName"
                to="TestPath/residents/1"
              />
            </div>
          </div>
        </div>
        <div
          className="lmc-card-footer"
        >
          <div
            className="lmc-flex-container"
          >
            <div
              id="intro-js-step-add-residents"
              style={
                Object {
                  "maxWidth": 160,
                }
              }
            >
              <GlyphButton
                block={true}
                color="success"
                glyph="plus"
                glyphStyle={Object {}}
                onClick={[Function]}
                position="left"
                title="Add a Resident"
              >
                <ResponsiveText
                  component="span"
                  visibleLG="Add a Resident"
                  visibleMD="Add a Resident"
                  visibleSM="Add a Resident"
                />
              </GlyphButton>
            </div>
            <div
              style={
                Object {
                  "maxWidth": 95,
                }
              }
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="TestPath/residents"
              >
                <Button
                  aphroditeStyles={Array []}
                  color="default"
                  variant="fill"
                >
                  <span
                    style={
                      Object {
                        "opacity": 0.6,
                      }
                    }
                  >
                    View All
                  </span>
                </Button>
              </Link>
            </div>
          </div>
        </div>
      </div>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <LmcResidentsCard
        onCreate={[Function]}
        residents={
          Array [
            Object {
              "id": 1,
              "name": "TestName",
              "status": "active",
            },
          ]
        }
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": LmcResidentsCard {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "onCreate": [Function],
          "residents": Array [
            Object {
              "id": 1,
              "name": "TestName",
              "status": "active",
            },
          ],
        },
        "refs": Object {},
        "renderResidents": [Function],
        "showMore": [Function],
        "state": Object {
          "max_residents_displayed": 10,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <h2
            className="lmc-card-title"
          >
            Residents
          </h2>
          <div
            className="lmc-card"
          >
            <div
              className="lmc-card-body"
            >
              <div>
                <div
                  className="lmc-flex-grid"
                >
                  <LmcProfileLink
                    name="TestName"
                    to="TestPath/residents/1"
                  />
                </div>
              </div>
            </div>
            <div
              className="lmc-card-footer"
            >
              <div
                className="lmc-flex-container"
              >
                <div
                  id="intro-js-step-add-residents"
                  style={
                    Object {
                      "maxWidth": 160,
                    }
                  }
                >
                  <GlyphButton
                    block={true}
                    color="success"
                    glyph="plus"
                    glyphStyle={Object {}}
                    onClick={[Function]}
                    position="left"
                    title="Add a Resident"
                  >
                    <ResponsiveText
                      component="span"
                      visibleLG="Add a Resident"
                      visibleMD="Add a Resident"
                      visibleSM="Add a Resident"
                    />
                  </GlyphButton>
                </div>
                <div
                  style={
                    Object {
                      "maxWidth": 95,
                    }
                  }
                >
                  <Link
                    onlyActiveOnIndex={false}
                    style={Object {}}
                    to="TestPath/residents"
                  >
                    <Button
                      aphroditeStyles={Array []}
                      color="default"
                      variant="fill"
                    >
                      <span
                        style={
                          Object {
                            "opacity": 0.6,
                          }
                        }
                      >
                        View All
                      </span>
                    </Button>
                  </Link>
                </div>
              </div>
            </div>
          </div>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <h2
            className="lmc-card-title"
          >
            Residents
          </h2>
          <div
            className="lmc-card"
          >
            <div
              className="lmc-card-body"
            >
              <div>
                <div
                  className="lmc-flex-grid"
                >
                  <LmcProfileLink
                    name="TestName"
                    to="TestPath/residents/1"
                  />
                </div>
              </div>
            </div>
            <div
              className="lmc-card-footer"
            >
              <div
                className="lmc-flex-container"
              >
                <div
                  id="intro-js-step-add-residents"
                  style={
                    Object {
                      "maxWidth": 160,
                    }
                  }
                >
                  <GlyphButton
                    block={true}
                    color="success"
                    glyph="plus"
                    glyphStyle={Object {}}
                    onClick={[Function]}
                    position="left"
                    title="Add a Resident"
                  >
                    <ResponsiveText
                      component="span"
                      visibleLG="Add a Resident"
                      visibleMD="Add a Resident"
                      visibleSM="Add a Resident"
                    />
                  </GlyphButton>
                </div>
                <div
                  style={
                    Object {
                      "maxWidth": 95,
                    }
                  }
                >
                  <Link
                    onlyActiveOnIndex={false}
                    style={Object {}}
                    to="TestPath/residents"
                  >
                    <Button
                      aphroditeStyles={Array []}
                      color="default"
                      variant="fill"
                    >
                      <span
                        style={
                          Object {
                            "opacity": 0.6,
                          }
                        }
                      >
                        View All
                      </span>
                    </Button>
                  </Link>
                </div>
              </div>
            </div>
          </div>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <LmcResidentsCard
    onCreate={[Function]}
    residents={
      Array [
        Object {
          "id": 1,
          "name": "TestName",
          "status": "active",
        },
      ]
    }
  />,
}
`;
